gadgets/trace_lsm/program.bpf.c

```diff
@@ -0,0 +1,222 @@
+	F(inode_follow_link)        \
+	F(inode_permission)         \
+	F(inode_setattr)            \
+	F(inode_getattr)            \
+	F(inode_setxattr)           \
+	F(inode_post_setxattr)      \
+	F(inode_getxattr)           \
+	F(inode_listxattr)          \
+	F(inode_removexattr)        \
+	F(inode_set_acl)            \
+	F(inode_get_acl)            \
+	F(inode_remove_acl)         \
+	F(inode_need_killpriv)      \
+	F(inode_killpriv)           \
+	F(inode_getsecurity)        \
+	F(inode_setsecurity)        \
+	F(inode_listsecurity)       \
+	F(inode_getsecid)           \
+	F(inode_copy_up)            \
+	F(inode_copy_up_xattr)      \
+	F(kernfs_init_security)     \
+	F(file_permission)          \
+	F(file_alloc_security)      \
+	F(file_free_security)       \
+	F(file_ioctl)               \
+	F(file_ioctl_compat)        \
+	F(mmap_addr)                \
+	F(mmap_file)                \
+	F(file_mprotect)            \
+	F(file_lock)                \
+	F(file_fcntl)               \
+	F(file_set_fowner)          \
+	F(file_send_sigiotask)      \
+	F(file_receive)             \
+	F(file_open)                \
+	F(file_truncate)            \
+	F(task_alloc)               \
+	F(task_free)                \
+	F(cred_alloc_blank)         \
+	F(cred_free)                \
+	F(cred_prepare)             \
+	F(cred_transfer)            \
+	F(cred_getsecid)            \
+	F(kernel_act_as)            \
+	F(kernel_create_files_as)   \
+	F(kernel_module_request)    \
+	F(kernel_load_data)         \
+	F(kernel_post_load_data)    \
+	F(kernel_read_file)         \
+	F(kernel_post_read_file)    \
+	F(task_fix_setuid)          \
+	F(task_fix_setgid)          \
+	F(task_fix_setgroups)       \
+	F(task_setpgid)             \
+	F(task_getpgid)             \
+	F(task_getsid)              \
+	F(current_getsecid_subj)    \
+	F(task_getsecid_obj)        \
+	F(task_setnice)             \
+	F(task_setioprio)           \
+	F(task_getioprio)           \
+	F(task_prlimit)             \
+	F(task_setrlimit)           \
+	F(task_setscheduler)        \
+	F(task_getscheduler)        \
+	F(task_movememory)          \
+	F(task_kill)                \
+	F(task_prctl)               \
+	F(task_to_inode)            \
+	F(userns_create)            \
+	F(ipc_permission)           \
+	F(ipc_getsecid)             \
+	F(msg_msg_alloc_security)   \
+	F(msg_msg_free_security)    \
+	F(msg_queue_alloc_security) \
+	F(msg_queue_free_security)  \
+	F(msg_queue_associate)      \
+	F(msg_queue_msgctl)         \
+	F(msg_queue_msgsnd)         \
+	F(msg_queue_msgrcv)         \
+	F(shm_alloc_security)       \
+	F(shm_free_security)        \
+	F(shm_associate)            \
+	F(shm_shmctl)               \
+	F(shm_shmat)                \
+	F(sem_alloc_security)       \
+	F(sem_free_security)        \
+	F(sem_associate)            \
+	F(sem_semctl)               \
+	F(sem_semop)                \
+	F(netlink_send)             \
+	F(d_instantiate)            \
+	F(getselfattr)              \
```

Comment:
On Linux 6.6.13-100.fc38.x86_64 I get:
```
program trace_lsm_getselfattr: attach LSM/LSMMac: getselfattr LSM hook not supported
```

This LSM hook was added in Linux 6.8:
https://github.com/torvalds/linux/commit/a04a1198088a1378d0389c250cc684f649bcc91e

For this PR, could you remove the hooks that were added recently, so this gadget works on most kernels?

---

I wonder how to make the user experience better (in another PR). What about a flag "besteffort":
```
SEC("lsm/getselfattr/besteffort")
```
And when defined, ig would ignore the error when the LSM hook is not supported in the current kernel.
In this way, the gadget author would decide if it's better to fail or not. Or maybe that should be something defined in gadget.yaml, so the ELF conventions don't diverge between libbpf, cilium/ebpf or ig. Alternatively, it could be something that the WASM module decide based on the BTF data, e.g. in `union security_list_options` (cc @mauriciovasquezbernal ). What do you think?

Reply:
The `besteffort` flag seems to be a general option. Maybe we can create a new macro `SEC_BESTEFFORT`, and it will add strings `[besteffort]` in the end of section names. We will identify and erase it while attaching.

Reply:
uprobe/libc:malloc?failstrategy=ignore&option2=foo2

Reply:
I'm wrong, the `cillum/ebpf` library will verify the section name while creating ebpf collections ([NewCollectionWithOptions](https://github.com/cilium/ebpf/blob/main/collection.go#L366)), by finding the name in kernel [findProgramTargetInKernel](https://github.com/cilium/ebpf/blob/main/prog.go#L1011). So if we add `?para&para2` in the end of section names, the ebpf collection will fail to load.

Reply:
For older kernels, the `attach LSM/LSMMac: getselfattr LSM hook not supported` error message also comes from `NewCollectionWithOptions`, not our attaching process. So to support the `besteffort` flag, we may also need to hack `cillum/ebpf`.

Reply:
Issue in cillium/ebpf: [#https://github.com/cilium/ebpf/discussions/1470](https://github.com/cilium/ebpf/discussions/1470)